lib/matplotlib/axes/_axes.py

```diff
@@ -5638,24 +5639,59 @@ def _pcolorargs(funcname, *args, allmatch=False):
             raise TypeError(
                 'Incompatible X, Y inputs to %s; see help(%s)' % (
                 funcname, funcname))
-        if allmatch:
-            if (Nx, Ny) != (ncols, nrows):
+
+        if shading == 'auto':
+            if ncols == Nx and nrows == Ny:
+                shading = 'nearest'
+            else:
+                shading = 'flat'
+
+        if shading == 'flat':
+            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-        else:
-            if not (ncols in (Nx, Nx - 1) and nrows in (Ny, Ny - 1)):
+            if (ncols == Nx or nrows == Ny):
+                cbook.warn_deprecated("3.3",
+                    message="shading='flat' when X and Y have the same "
+                            "dimensions as C is deprecated since %(since)s.  "
+                            "Either specify the corners of the quadrilaterals "
+                            "with X and Y, or pass shading='auto', 'nearest' "
+                            "or 'gouraud', or set rcParams['pcolor.shading']")
+            C = C[:Ny - 1, :Nx - 1]
+        else:    # ['nearest', 'gouraud']:
+            if (Nx, Ny) != (ncols, nrows):
                 raise TypeError('Dimensions of C %s are incompatible with'
                                 ' X (%d) and/or Y (%d); see help(%s)' % (
                                     C.shape, Nx, Ny, funcname))
-            C = C[:Ny - 1, :Nx - 1]
+            if shading in ['nearest', 'auto']:
+                # grid is specified at the center, so define corners
+                # at the midpoints between the grid centers and then use the
+                # flat algorithm.
+                def _interp_grid(X):
+                    # helper for below
+                    dX = np.diff(X, axis=1)/2.
+                    X = np.hstack((X[:, [0]] - dX[:, [0]],
```

Comment:
Symmetrically expanding the first and last pixels seem fine but needs to be documented below.  Also I guess this will crash if X or Y has size 1?  (Sure it's degenerate, but I guess you can either use a width of 1 (likewise needs doc), or error out more cleanly.)

Reply:
I guess we should error.  Assuming width 1 seems undesirable....

Reply:
erroring is certainly fine with me

Reply:
I think the general mpl strategy, though, is to try hard not to error out when data are simply missing:
```python
In [2]: plt.plot([])
Out[2]: [<matplotlib.lines.Line2D at 0x119b13898>]
```
The idea is that if you have a long-running process making plots from a stream of data that might have gaps, it is better to show an empty window during a gap than to require that the user trap the gap condition and handle it manually.

Reply:
OK, fair enough.  Instead of guessing a dX, I just make it [X, X] and nothing visible gets plotted.  Anything else seems liable to just be wrong (dX = X/2, has problems if X=0, and just choosing dX=1/2 seems like it could mess with scaling of some plots inadvertently.